File: ios/Classes/TiGooglemapsAutocompleteDialogProxy.m
/**
 * Appcelerator Titanium Mobile
 * Copyright (c) 2009-2016 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

#import "TiGooglemapsAutocompleteDialogProxy.h"
#import "TiApp.h"

@implementation TiGooglemapsAutocompleteDialogProxy

#pragma mark Memory Management

- (void)dealloc
{
    dialog.delegate = nil;
    RELEASE_TO_NIL(dialog);
    
    [super dealloc];
}

#pragma mark Public APIs

- (GMSAutocompleteViewController*)dialog
{
    if (dialog == nil) {
        dialog = [GMSAutocompleteViewController new];
        [dialog setDelegate:self];
    }
    
    return dialog;
}

- (void)configure:(id)value
{
    ENSURE_SINGLE_ARG(value, NSString);
    [GMSPlacesClient provideAPIKey:value];
}

- (void)open:(id)args
{
    [self rememberSelf];
    ENSURE_UI_THREAD(open, args);

    id animated = [args valueForKey:@"animated"];
    ENSURE_TYPE_OR_NIL(animated, NSNumber);
    
    [[TiApp app] showModalController:[self dialog] animated:[TiUtils boolValue:animated def:YES]];
}

- (void)setTableCellBackgroundColor:(id)value
{
    ENSURE_TYPE(value, NSString);
    [[self dialog] setTableCellBackgroundColor:[[TiUtils colorValue:value] color]];
}

- (void)setTableCellSeparatorColor:(id)value
{
    ENSURE_TYPE(value, NSString);
    [[self dialog] setTableCellSeparatorColor:[[TiUtils colorValue:value] color]];
}

- (void)setPrimaryTextColor:(id)value
{
    ENSURE_TYPE(value, NSString);
    [[self dialog] setPrimaryTextColor:[[TiUtils colorValue:value] color]];
}

- (void)setPrimaryTextHighlightColor:(id)value
{
    ENSURE_TYPE(value, NSString);
    [[self dialog] setPrimaryTextHighlightColor:[[TiUtils colorValue:value] color]];
}

- (void)setSecondaryTextColor:(id)value
{
    ENSURE_TYPE(value, NSString);
    [[self dialog] setSecondaryTextColor:[[TiUtils colorValue:value] color]];
}

- (void)setTintColor:(id)value
{
    ENSURE_TYPE(value, NSString);
    [[self dialog] setTintColor:[[TiUtils colorValue:value] color]];
}

#pragma mark - Delegates

- (void)viewController:(GMSAutocompleteViewController *)viewController didAutocompleteWithPlace:(GMSPlace *)place
{
    if ([self _hasListeners:@"success"]) {
        [self fireEvent:@"success" withObject:@{
            @"place": [TiGooglemapsAutocompleteDialogProxy dictionaryFromPlace:place]
        }];
    }
    
    [self closeDialog];
}

- (void)viewController:(GMSAutocompleteViewController *)viewController didFailAutocompleteWithError:(NSError *)error
{
    if ([self _hasListeners:@"error"]) {
        [self fireEvent:@"error" withObject:@{
            @"error": [error localizedDescription],
            @"code": NUMINTEGER([error code])
        }];
    }
    
    [self closeDialog];
}

- (void)wasCancelled:(GMSAutocompleteViewController *)viewController
{    
    if ([self _hasListeners:@"cancel"]) {
        [self fireEvent:@"cancel" withObject:nil];
    }
    
    [self closeDialog];
}

#pragma mark Utilities

- (void)closeDialog
{
    [[self dialog] setDelegate:nil];
    [[self dialog] dismissViewControllerAnimated:YES completion:nil];
    [self forgetSelf];
    RELEASE_TO_NIL(dialog);
}

+ (NSDictionary*)dictionaryFromPlace:(GMSPlace*)place
{
    return @{
        @"name": [place name],
        @"placeID": [place placeID],
        @"latitude": NUMDOUBLE([place coordinate].latitude),
        @"longitude": NUMDOUBLE([place coordinate].longitude),
        @"formattedAddress": [place formattedAddress],
        @"addressComponents": [TiGooglemapsAutocompleteDialogProxy arrayFromAddressComponents:[place addressComponents]]
    };
}

+ (id)arrayFromAddressComponents:(NSArray<GMSAddressComponent*>*)addressComponents
{
    if (addressComponents == nil) {
        return [NSNull null];
    }
    
    NSMutableArray *result = [NSMutableArray new];
    for (GMSAddressComponent *addressComponent in addressComponents) {
        [result addObject:@{
            @"type": addressComponent.type,
            @"name":addressComponent.name
        }];
    }
    
    return [result autorelease];
}

+ (id)dictionaryFromPrediction:(GMSAutocompletePrediction*)prediction
{
    if (prediction == nil) {
        return nil;
    }
    
    return @{
        @"attributedFullText": [[prediction attributedFullText] string],
        @"attributedPrimaryText": [[prediction attributedPrimaryText] string],
        @"attributedSecondaryText": [[prediction attributedSecondaryText] string],
        @"placeID": [prediction placeID],
        @"types": [prediction types]
    };
}

@end

File: titanium-googlemaps-swift/Classes/TiGooglemapsAutocompleteDialogProxy.swift
/**
 * Axway Titanium
 */

import UIKit
import TitaniumKit
import GooglePlaces

@objc(TiGooglemapsAutocompleteDialogProxy)
public class TiGooglemapsAutocompleteDialogProxy: TiProxy, GMSAutocompleteViewControllerDelegate {
  private var autocompleteController: GMSAutocompleteViewController?
  
  private func controller() -> GMSAutocompleteViewController {
    if let controller = autocompleteController {
      return controller
    }
    
    let controller = GMSAutocompleteViewController()
    controller.delegate = self
    autocompleteController = controller
    return controller
  }
  
  private func resetController() {
    autocompleteController?.delegate = nil
    autocompleteController = nil
  }
  
  // MARK: - Public API
  
  @objc public func configure(_ value: String) {
    GMSPlacesClient.provideAPIKey(value)
  }
  
  @objc public func open(_ args: [String: Any]) {
    rememberSelf()
    
    runOnMainThread {
      let animated = TiUtils.boolValue(args["animated"], def: true)
      TiApp.sharedApp().showModalController(self.controller(), animated: animated)
    }
  }
  
  @objc public func setTableCellBackgroundColor(_ value: Any?) {
    controller().tableCellBackgroundColor = TiUtils.colorValue(value).color
  }
  
  @objc public func setTableCellSeparatorColor(_ value: Any?) {
    controller().tableCellSeparatorColor = TiUtils.colorValue(value).color
  }
  
  @objc public func setPrimaryTextColor(_ value: Any?) {
    controller().primaryTextColor = TiUtils.colorValue(value).color
  }
  
  @objc public func setPrimaryTextHighlightColor(_ value: Any?) {
    controller().primaryTextHighlightColor = TiUtils.colorValue(value).color
  }
  
  @objc public func setSecondaryTextColor(_ value: Any?) {
    controller().secondaryTextColor = TiUtils.colorValue(value).color
  }
  
  @objc public func setTintColor(_ value: Any?) {
    controller().tintColor = TiUtils.colorValue(value).color
  }
  
  // MARK: - Delegate
  
  public func viewController(_ viewController: GMSAutocompleteViewController, didAutocompleteWith place: GMSPlace) {
    if _hasListeners("success") {
      fireEvent("success", with: ["place": Self.dictionary(from: place)])
    }
    
    closeDialog()
  }
  
  public func viewController(_ viewController: GMSAutocompleteViewController, didFailAutocompleteWithError error: Error) {
    if _hasListeners("error") {
      let nsError = error as NSError
      fireEvent("error", with: ["error": error.localizedDescription, "code": nsError.code])
    }
    
    closeDialog()
  }
  
  public func wasCancelled(_ viewController: GMSAutocompleteViewController) {
    if _hasListeners("cancel") {
      fireEvent("cancel", with: nil)
    }
    
    closeDialog()
  }
  
  // MARK: - Utilities
  
  private func runOnMainThread(_ block: @escaping () -> Void) {
    if Thread.isMainThread {
      block()
    } else {
      TiThreadPerformOnMainThread(block, false)
    }
  }
  
  private func closeDialog() {
    runOnMainThread {
      self.autocompleteController?.dismiss(animated: true) {
        self.resetController()
        self.forgetSelf()
      }
    }
  }
  
  @objc public class func dictionary(from place: GMSPlace) -> [String: Any] {
    return [
      "name": place.name ?? NSNull(),
      "placeID": place.placeID ?? NSNull(),
      "latitude": place.coordinate.latitude,
      "longitude": place.coordinate.longitude,
      "formattedAddress": place.formattedAddress ?? NSNull(),
      "addressComponents": array(from: place.addressComponents)
    ]
  }
  
  @objc public class func array(from components: [GMSAddressComponent]?) -> Any {
    guard let components = components else {
      return NSNull()
    }
    
    return components.map {
      [
        "types": $0.types,
        "name": $0.name
      ]
    }
  }
  
  /*
  @objc public class func dictionary(from prediction: GMSAutocompletePrediction?) -> [String: Any]? {
    guard let prediction = prediction else { return nil }
    
    return [
      "attributedFullText": prediction.attributedFullText.string,
      "attributedPrimaryText": prediction.attributedPrimaryText.string,
      "attributedSecondaryText": prediction.attributedSecondaryText?.string as Any,
      "placeID": prediction.placeID ?? NSNull(),
      "types": prediction.types
    ]
  }*/
}
